p4_symbolic/ir/ir.proto

```diff
@@ -0,0 +1,106 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The main protobuf file for definition of our intermediate representation.
+// This file defines the overall structure of a P4 program in our IR.
+// The structure is tailored to be easy to consume rather than produce.
+// It deviates from the underlying bmv2 JSON format, which is mimiced closely
+// by p4_symbolic/bmv2/bmv2.proto.
+// The bmv2 proto is used to initially parse the JSON format, which is then
+// processed by our tool and transformed into the IR.
+// The IR does not cover all the features of P4 programs.
+// Specifically, we intentionally ignore parsers, de-parsers, stacks and unions,
+// checksum calculations, and extern definitions.
+// Additionally, some types of expressions and statements are unsupported, these
+// are described in values.proto and statements.proto respectively.
+
+syntax = "proto3";
+
+package p4_symbolic.ir;
+
+import "p4_pdpi/ir.proto";
+import "p4_symbolic/bmv2/bmv2.proto";
+import "p4_symbolic/ir/statements.proto";
+
+// Overall structure of a p4 program.
+message P4Program {
+  // Header definitions, keyed by the header type name (e.g. standard_metadata).
+  map<string, HeaderType> headers = 1;
+  // Action definitions and implementations, keyed by action name.
+  map<string, Action> actions = 2;
+  // Table definitions and implementations, keyed by table name.
+  map<string, Table> tables = 3;
+  // The name of the initial table.
+  string initial_table = 6;
```

Comment:
Is this an `initial_table` a standard P4 thing? I haven't heard about this before.

Reply:
Yes, Stefan also asked about it.
This is the first table in the first pipeline. You can see in the JSON format doc that every pipeline has one.
https://github.com/p4lang/behavioral-model/blob/master/docs/JSON_format.md#pipelines 

In my sample programs, I noticed that even programs without table (e.g. hardcoded, where routing is based on a hardcoded if statement) do end up having an initial table, since the if statement is itself transformed into a table. This seems true for programs where the table to match on is also determined dynamically via an if statement etc.

This is pretty useful here, because it immediately tells me where to start the symbolic processing. It is in the bmv2 format, so I do not even have to find it myself.

Reply:
Will we need it? Seems like this information should be given by apply block of each pipeline? If so, should we remove it?